### ISSUE TITLE
Fatal: EOF when optimizing dottyguide.pdf

### ISSUE THREAD

**Issue by behrangsa** [hhrutter/pdfcpu#86]
Error:

```
./pdfcpu optimize -v -stats foo.csv /tmp/pdf/dottyguide.pdf 
stats will be appended to foo.csv
 INFO: 2019/06/06 16:22:52 reading /tmp/pdf/dottyguide.pdf..
 INFO: 2019/06/06 16:22:52 PDF Version 1.5 conforming reader
Fatal: EOF
readXRefTable: unexpected eof
github.com/hhrutter/pdfcpu/pkg/pdfcpu.readXRefTable
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/read.go:1110
github.com/hhrutter/pdfcpu/pkg/pdfcpu.Read
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/read.go:78
github.com/hhrutter/pdfcpu/pkg/pdfcpu.ReadFile
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/read.go:57
github.com/hhrutter/pdfcpu/pkg/api.ReadContextFromFile
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:127
github.com/hhrutter/pdfcpu/pkg/api.readAndValidate
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:236
github.com/hhrutter/pdfcpu/pkg/api.readValidateAndOptimize
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:255
github.com/hhrutter/pdfcpu/pkg/api.Optimize
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:287
github.com/hhrutter/pdfcpu/pkg/api.Process
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/process.go:84
main.process
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:219
main.main
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:182
runtime.main
	/usr/lib/go-1.10/src/runtime/proc.go:198
runtime.goexit
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361
Read: xRefTable failed
github.com/hhrutter/pdfcpu/pkg/pdfcpu.Read
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/read.go:80
github.com/hhrutter/pdfcpu/pkg/pdfcpu.ReadFile
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/pdfcpu/read.go:57
github.com/hhrutter/pdfcpu/pkg/api.ReadContextFromFile
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:127
github.com/hhrutter/pdfcpu/pkg/api.readAndValidate
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:236
github.com/hhrutter/pdfcpu/pkg/api.readValidateAndOptimize
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:255
github.com/hhrutter/pdfcpu/pkg/api.Optimize
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/api.go:287
github.com/hhrutter/pdfcpu/pkg/api.Process
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/pkg/api/process.go:84
main.process
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:219
main.main
	/home/buntuu/go/src/github.com/hhrutter/pdfcpu/cmd/pdfcpu/main.go:182
runtime.main
	/usr/lib/go-1.10/src/runtime/proc.go:198
runtime.goexit
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361
```

Please refer to the attached file.

---

[dottyguide.pdf](https://github.com/hhrutter/pdfcpu/files/3260236/dottyguide.pdf)

**Comment by hhrutter** [comment]
This file has corrupt object file pointers.
Working on a fix..

**Comment by hhrutter** [comment]
Sorrry, I've been looking into it but at this point pdfcpu cannot repair a corrupt xreftable.
The flexible validation feature is meant to be for out of spec anomalies but still relying on a fine xRefTable. The whole reader would have to be rewritten in order to digest such files.

Your best bet is to eg. open the file with Mac Preview then Export to PDF.
Then do a `pdfcpu optimize yourfile.pdf` and process the result file with pdfcpu in any way you like.

Thanks for using pdfcpu!

**Comment by behrangsa** [comment]
All good! Thanks for looking into it.

PS: It seems to me that pdfcpu cannot optimize/deduplicate Type 3 fonts at
the moment, is that correct?

Best regards,
Behrang Saeedzadeh
blog.behrang.org

**Comment by hhrutter** [comment]
This is fixed with commit [86b3445](https://github.com/hhrutter/pdfcpu/commit/86b344515ab4f12e80ea26f3c0e519e28047274f).

Optimization works at the image object and font object level.
It does not go into parsing and logical comparing the content of font files for redundancies on
a char or glyph level in case for Type 3 fonts.

**Comment by behrangsa** [comment]
Perfect. Thanks!
